### PR TITLE
feat: Implement barebones Forum channel support.

### PR DESCRIPTION
**Pull request by FayeDel**
## About

This PR title essentially explains itself.
Forums are in testing but has been more widespread, so this is left as a draft PR until its live/more features are discovered.

This correlates to an earlier PR: #630 

## Checklist

- [X] I've ran `pre-commit` to format and lint the change(s) made.
- [X] I've checked to make sure the change(s) work on `3.8.6` and higher.
- [ ] This fixes/solves an [Issue](https://github.com/goverfl0w/discord-interactions/issues).
  - (If existent):
- [X] I've made this pull request for/as: (check all that apply)
  - [X] Documentation
  - [X] Breaking change
  - [X] New feature/enhancement
  - [ ] Bugfix

### CODE REVIEW

**Comment by i0bs** [comment]
As of 5 July.

image

File: interactions/api/models/channel.py [hunk 1]
```diff
@@ -1201,10 +1228,60 @@ async def join(self) -> None:
...
+
+    ...
+
+
+@define()
+class Forum(Thread):
+    """An object representing a Forum.
+    .. note::
+        While this isn't directly dispatched for GW based events,
+        this object exists moreso for future helper methods.
```
Comment (sillyElvy) [inline interactions/api/models/channel.py]: Wouldn't it make more sense to keep the methods on the channel object like we did for threads?

**Comment by FayeDel** [comment]
Forums are officially out so I'm marking this as ready for review ^^

File: interactions/api/models/channel.py [hunk 2]
```diff
@@ -95,6 +96,22 @@ class ThreadMember(ClientSerializerMixin):
...
+    #     "flags",
+    #     # TODO: Document below attributes.
+    #     "muted",  # ??
+    #     "user",
+    #     "team_id",
+    #     "membership_state",
+    #     "permissions",
+    #     "member",
+    #     "presence",
+    # )
```
Comment (sillyElvy) [inline interactions/api/models/channel.py]: _uh_

File: interactions/api/models/channel.py [hunk 3]
```diff
@@ -104,6 +121,12 @@ class ThreadMember(ClientSerializerMixin):
         default=None, repr=False
     )  # todo explore this, it isn't in the ddev docs
 
+    # Take these conversions with a grain of salt.
+
+    # Commented out for circular import reasons, but schema says they do work.
+    # self.member = Member(**self._json.get("member")) if self._json.get("member") else None
+    # self.presence = Presence(**self._json.get("presence")) if self._json.get("presence") else None
```
Comment (sillyElvy) [inline interactions/api/models/channel.py]: Will you do something with this or just let it exist?

File: interactions/api/models/channel.py [hunk 4]
```diff
@@ -319,6 +342,15 @@ class Channel(ClientSerializerMixin, IDMixin):
...
     """
 
+    # __slots__ = (
+    #     # TODO: Document banner when Discord officially documents them.
+    #     "banner",
+    #     "guild_hashes",
+    #     "flags",
+    #     "template",
+    #     "available_tags",
+    # )
```
Comment (sillyElvy) [inline interactions/api/models/channel.py]: _uh_

File: interactions/api/models/channel.py [hunk 5]
```diff
@@ -1505,10 +1537,60 @@ async def get_permissions_for(self, member: "Member") -> Permissions:
...
+    #     # TODO: Document these when Discord officially documents them.
+    #     "banner",
+    #     "guild_hashes",
+    #     "flags",  # This seems to be forum unique.
+    #     "newly_created",
+    #     "template",
+    #     "available_tags",
+    # )
+
+    ...
```
Comment (sillyElvy) [inline interactions/api/models/channel.py]: _uh_

File: interactions/api/models/channel.py [hunk 6]
```diff
@@ -1505,10 +1537,60 @@ async def get_permissions_for(self, member: "Member") -> Permissions:
...
+
+
+@define()
+class Forum(Thread):
+    """An object representing a Forum.
+    .. note::
+        While this isn't directly dispatched for GW based events,
+        this object exists moreso for future helper methods.
```
Comment (sillyElvy) [inline interactions/api/models/channel.py]: Helpers are generally implemented in the base channel class so people don't have to use different objects for different channels - if this is not needed, it should be removed

File: interactions/api/models/gw.py [hunk 7]
```diff
@@ -517,6 +517,10 @@ class MessageReactionRemove(MessageReaction):
     # todo see if the missing member attribute affects anything
 
 
+# TODO: (temp) Add Thread object for a THREAD_UPDATE event.
+# More prevalent on forums for some odd reason.
```
Comment (sillyElvy) [inline interactions/api/models/gw.py]: If its forum related why not add it?

**Review by sillyElvy — CHANGES_REQUESTED** [review]

**Comment by sillyElvy** [comment]
Docs build failed

**Review by sillyElvy — APPROVED** [review]
Lgtm!

**Pull request merged**